src/tactic/local_cache.lean

```diff
@@ -0,0 +1,211 @@
+import tactic.basic
+import init.data.nat.bitwise
+
+namespace tactic
+namespace local_cache
+
+section internal
+
+variables {α : Type} [reflected α] [has_reflect α]
+
+meta def mk_full_namespace (ns : name) : name := `local_cache ++ ns
+
+meta def save_data (dn : name) (a : α) [reflected a] : tactic unit :=
+tactic.add_decl $ mk_definition dn [] (reflect α) (reflect a)
+
+meta def load_data (dn : name) : tactic α :=
+do e ← tactic.get_env,
+   d ← e.get dn,
+   tactic.eval_expr α d.value
+
+meta def poke_data (dn : name) : tactic bool :=
+do e ← tactic.get_env,
+   return (e.get dn).to_bool
+
+meta def run_once_under_name {α : Type} [reflected α] [has_reflect α] (t : tactic α) (cache_name : name) : tactic α :=
+do load_data cache_name <|>
+   do {
+     a ← t,
+     save_data cache_name a,
+     return a
+   }
+
+end internal
+
+-- We maintain two separate caches with different scopes:
+-- one local to `begin ... end` or `by` blocks, and another
+-- for entire `def`/`lemma`s.
+meta structure cache_scope :=
+-- Returns the name of the def used to store the contents of is cache,
+-- making a new one and recording this in private state if neccesary.
+(get_name : name → tactic name)
+-- Same as above but fails instead of making a new name, and never
+-- mutates state.
+(try_get_name : name → tactic name)
+-- Asks whether the namespace `ns` currently has a value-in-cache
+(present : name → tactic bool)
+-- Clear cache associated to namespace `ns`
+(clear : name → tactic unit)
+
+namespace block_local
+
+-- `mk_new` gives a way to generate a new name if no current one
+-- exists.
+private meta def get_name_aux (ns : name) (mk_new : options → name → tactic name) : tactic name :=
+do o ← tactic.get_options,
+   let opt := mk_full_namespace ns,
+   match o.get_string opt "" with
+   | "" := mk_new o opt
+   | s := return $ name.from_components $ s.split (= '.')
+   end
+
+meta def get_name (ns : name) : tactic name :=
+get_name_aux ns $ λ o opt,
+do n ← mk_user_fresh_name,
+   tactic.set_options $ o.set_string opt n.to_string,
+   return n
+
+-- Like `get_name`, but fail if `ns` does not have a cached
+-- decl name (we create a new one above).
+meta def try_get_name (ns : name) : tactic name :=
+get_name_aux ns $ λ o opt, fail format!"no cache for \"{ns}\""
+
+meta def present (ns : name) : tactic bool :=
+do o ← tactic.get_options,
+   match o.get_string (mk_full_namespace ns) "" with
+   | "" := return ff
+   | s  := return tt
+   end
+
+meta def clear (ns : name) : tactic unit :=
+do o ← tactic.get_options,
+   set_options $ o.set_string (mk_full_namespace ns) ""
+
+end block_local
+
+namespace def_local
+
+-- Fowler–Noll–Vo hash function (FNV-1a)
+section fnv_a1
+
+def FNV_OFFSET_BASIS := 0xcbf29ce484222325
+def FNV_PRIME := 0x100000001b3
+def RADIX := 0xFFFFFFFFFFFFFFFF + 1 -- 64-bits
```

Comment:
You could use norm_num here so that it's self documenting: 

```lean
def RADIX := (by apply_normed (2^64-1)) + 1 -- 64-bits
```

Reply:
The only reason I did it this way was so the number of bits you got back was manifest---just writing `2^64` is fine too, so I guess I'll do that if you prefer.